### PR TITLE
GPIO driver for the H3 SoC based boards (OrangePi, NanoPi)

### PR DESCRIPTION
**Pull request by MX-Master**
Driver usage guide - [H3_GPIO_driver_usage.pdf](http://topcnc.ru/opi1_cnc/5%20-%20Machinekit,%20LinuxCNC%20and%20GPIO.pdf)

Tested with
- **Orange Pi One**
- **Orange Pi Lite**
- **Orange Pi PC**

### CODE REVIEW

**Comment by ArcEye** [comment]
Author not whitelisted yet - can one of the admins verify this patch?

**Comment by l29ah** [comment]
The driver can be modified to accommodate all the Allwinner SoCs by purging the hardcoded pin table.

**Comment by ArcEye** [comment]
ok to test

**Comment by MX-Master** [comment]
I have some files for those people who wants to make a quick test of **GPIO** driver with **Orange Pi One** board. I made two **MicroSD** images (8Gb, 16Gb) with a couple of **HOWTO** files. **MicroSD** image - it's **Debian Jessie** (built by Armbian) with real-time kernel (4.13, RT-PREEMPT), with **Machinekit** and some tweaks installed:

- [opi1_machinekit_8G.rar](http://topcnc.ru/opi1_cnc/opi1_machinekit_8G.rar) (1.2Gb)
- [opi1_machinekit_16G.rar](http://topcnc.ru/opi1_cnc/opi1_machinekit_16G.rar) (1.2Gb)

**HOWTO**s:

1. [Hardware.pdf](http://topcnc.ru/opi1_cnc/1%20-%20Hardware.pdf)
2. [copy SD image to SD card.pdf](http://topcnc.ru/opi1_cnc/2%20-%20copy%20SD%20image%20to%20SD%20card.pdf)
3. [Control.pdf](http://topcnc.ru/opi1_cnc/3%20-%20Control.pdf)
4. [Remote Control.pdf](http://topcnc.ru/opi1_cnc/4%20-%20Remote%20Control.pdf)
5. [Machinekit, LinuxCNC and GPIO.pdf](http://topcnc.ru/opi1_cnc/5%20-%20Machinekit,%20LinuxCNC%20and%20GPIO.pdf)
6. [Wiring.pdf](http://topcnc.ru/opi1_cnc/6%20-%20Wiring.pdf)

**Comment by ArcEye** [comment]
As it is a new driver module, it will need to be specifically built.
I will merge it in a while, but it will not be built, until the Makefiles, configuration and packaging are amended.

I will look at that separately.
(The hal_gpio_demo.hal also needs `fp` adding the the servo thread line, has done since February, just shows no-one uses it 😄 )

You have included a lot of useful information, which is effectively going to be lost once this PR is merged.
Can I suggest for now, that you include a `hal_gpio_h3.readme` file which contains all those links and can sit next to it in `src/hal/drivers`.
Then it will be accessible at a fixed location and could be brought into the docs in due course.

**Comment by MX-Master** [comment]
OK. I will make some changes in the **Makefile** and **hal_gpio_h3.c** files. And will add the **hal_gpio_h3.readme**.

**Comment by ArcEye** [comment]
You can leave the Makefile to me if you want.
It will require configure.ac to be changed to introduce a new configure switch in addition to the build rule, plus the packaging rules will need checking and possibly updating to specifically include for arm.

**Comment by MX-Master** [comment]
I made a few changes to **Makefile**, **configure.ac** and **hal_gpio_h3.c**.
Please check the latest commit. Perhaps I missed something.

**Comment by ArcEye** [comment]
OK, I am out until later today, 
Looks about right, will build your branch in an armhf docker image to check it.

File: src/Makefile [hunk 1]
```diff
@@ -1406,6 +1406,11 @@ obj-m += hal_arm335xQEP.o
 hal_arm335xQEP-objs := hal/drivers/hal_arm335xQEP.o $(MATHSTUB)
 endif
 
+ifdef TARGET_PLATFORM_H3
+obj-$(CONFIG_HAL_GPIO) += hal_gpio_h3.o
+hal_gpio-objs := hal/drivers/hal_gpio_h3.o
```
Comment (ArcEye) [inline src/Makefile]: `hal_gpio-objs` already exists, should be `hal_gpio_h3-objs`

File: src/Makefile [hunk 2]
```diff
@@ -1849,6 +1854,10 @@ $(RTLIBDIR)/hal_arm335xQEP$(MODULE_EXT): $(addprefix $(OBJDIR)/,$(hal_arm335xQEP
 endif
 endif
 
+ifdef TARGET_PLATFORM_H3
+$(RTLIBDIR)/hal_gpio_h3$(MODULE_EXT): $(addprefix $(OBJDIR)/,$(hal_gpio-objs))
```
Comment (ArcEye) [inline src/Makefile]: .. and also here should be `hal_gpio_h3-objs`

**Comment by ArcEye** [comment]
The module does not get built.
I made the changes as per the notes I made on the code, but is only being built if I added it to the existing TARGET_PLATFORM_RASPBERRY rules

This presumably means that TARGET_PLATFORM_H3 is not getting defined despite the configure.log
showing
```
configure:5233: checking platform-h3
configure:5235: result: default enabled for this architecture

TARGET_PLATFORM_BEAGLEBONE='yes'
TARGET_PLATFORM_CHIP='yes'
TARGET_PLATFORM_H3='yes'
TARGET_PLATFORM_PC=''
TARGET_PLATFORM_RASPBERRY='yes'
TARGET_PLATFORM_SOCFPGA='yes'
TARGET_PLATFORM_ZEDBOARD='yes'
```
There probably a typo somewhere, and underscore used instead of a dash or similar, but it is not leaping out at me.
Look again tomorrow.

**Comment by MX-Master** [comment]
Fixed noted typos and made a small try to make module building more successful. **ArcEye**, if you want you can make changes directly to **hal_gpio_h3** branch - https://github.com/MX-Master/machinekit/invitations

**Comment by ArcEye** [comment]
The problem is as I suggested, `TARGET_PLATFORM_H3` is not getting defined in the Makefile despite it showing as being defined in `configure`.

When I inserted a test print into each `#ifdef TARGET_PLATFORM_XXX` test in the Makefile, the print does not include H3
```
"TARGET_PLATFORM_RASPBERRY set and building"
"TARGET_PLATFORM_SOCFPGA set and building"
"TARGET_PLATFORM_BEAGLEBONE set and building"
"TARGET_PLATFORM_CHIP set and building"
"TARGET_PLATFORM_ZEDBOARD set and building
```
Why is harder to understand, since there is a specific AC_SUBST() instruction in configure.ac

I'll keep looking

**Comment by ArcEye** [comment]
OK, fixed it, there was no corresponding entry in `Makefile.inc.in` to transfer the value.
I'll push something later.

**Comment by ArcEye** [comment]
Currently doing a package build inside a chroot armhf partition, to check that everything is as it should be and the driver goes into armhf packages.  Will take 30 mins or so.

**Comment by MX-Master** [comment]
Nice work, thanks

**Comment by ArcEye** [comment]
That all builds OK.
Thanks for your contribution.

**Pull request merged**